plugins/MachineSettingsAction/MachineSettingsAction.py

```diff
@@ -21,27 +21,56 @@ def __init__(self, parent = None):
 
         self._container_index = 0
 
-        cura.Settings.CuraContainerRegistry.getInstance().containerAdded.connect(self._onContainerAdded)
+        self._container_registry = UM.Settings.ContainerRegistry.getInstance()
+        self._container_registry.containerAdded.connect(self._onContainerAdded)
 
     def _reset(self):
         global_container_stack = UM.Application.getInstance().getGlobalContainerStack()
-        if global_container_stack:
+        if not global_container_stack:
+            return
+
+        # First check if there is a variant previously generated by this machine
+        machine_settings_variant = global_container_stack.findContainer({"type": "variant", "subtype": "machine_settings"})
+        if not machine_settings_variant:
+            # There may be a variant created by the UMOUpgradeSelection machine action
+            machine_settings_variant = global_container_stack.findContainer({"type": "variant", "id": global_container_stack.getName() + "_variant"})
+
+        if not machine_settings_variant:
             variant = global_container_stack.findContainer({"type": "variant"})
-            if variant:
-                variant_index = global_container_stack.getContainerIndex(variant)
-                if variant_index != self._container_index:
-                    self._container_index = variant_index
-                    self.containerIndexChanged.emit()
-                if variant.getId() == "empty_variant":
-                    self._createVariant(global_container_stack, self._container_index)
-
-    def _createVariant(self, global_container_stack):
-        # Create and switch to a variant to store the settings in
-        new_variant = UM.Settings.InstanceContainer(global_container_stack.getName() + "_variant")
-        new_variant.addMetaDataEntry("type", "variant")
-        new_variant.setDefinition(global_container_stack.getBottom())
-        UM.Settings.ContainerRegistry.getInstance().addContainer(new_variant)
-        global_container_stack.replaceContainer(self._container_index, new_variant)
+            if variant and variant.getId() == "empty_variant":
+                # There is an empty variant that we can use to store the machine settings
+                container_index = global_container_stack.getContainerIndex(variant)
+                machine_settings_variant = self._createMachineVariant(global_container_stack, container_index)
+            else:
+                # Add a second variant before the current variant to store the machine settings
+                machine_settings_variant = self._createMachineVariant(global_container_stack)
```

Comment:
The default way of finding a specific profile (e.g. the material) is by asking the container stack for the first profile with a specific type, like we do in line 39 here. I fear that that would all break if there are two containers with the same type in one stack. That makes me hesitatant to merge this.